pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexType.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.index;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.creator.IndexCreationContext;
+import org.apache.pinot.segment.spi.index.column.ColumnIndexContainer;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.IndexConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+/**
+ * TODO: implement mutable indexes.
+ * @param <C> the class that represents how this object is configured.
+ * @param <IR> the {@link IndexReader} subclass that should be used to read indexes of this type.
+ * @param <IC> the {@link IndexCreator} subclass that should be used to create indexes of this type.
+ */
+public interface IndexType<C extends IndexConfig, IR extends IndexReader, IC extends IndexCreator> {
+
+  /**
+   * The unique id that identifies this index type.
+   * <p>The returned value for each index should be constant across different Pinot versions as it is used as:</p>
+   *
+   * <ul>
+   *   <li>They key used when the index is registered in IndexService.</li>
+   *   <li>The internal identification in v1 files and metadata persisted on disk.</li>
+   *   <li>The default toString implementation.</li>
+   *   <li>The key that identifies the index config in the indexes section inside
+   *   {@link org.apache.pinot.spi.config.table.FieldConfig}, although specific index types may choose to read other
+   *   names (for example, <code>inverted_index</code> may read <code>inverted</code> key.</li>
+   * </ul>
+   */
+  String getId();
+
+  Class<C> getIndexConfigClass();
+
+  /**
+   * The default config when it is not explicitly defined by the user.
+   */
+  C getDefaultConfig();
+
+  C getConfig(TableConfig tableConfig, Schema schema);
+
+  /**
+   * Transforms a config object into a Jackson {@link JsonNode}.
+   */
+  default JsonNode serialize(C config) {
```

Comment:
`IndexConfig` already extends `BaseJsonConfig` which has `toJsonNode()` method. Do you see the need for the customized serializer?

Reply:
I think it is better to delegate this things into the one that actually defines how to use the object instead to the object itself. Which is related to whether to prefer OOP vs a more functional one with traits, extensions and stuff. But I don't expect to have indexes that need to override the serialization soon, so I'm going to remove the method. Also, the method was added before config extended BaseJsonConfig, so it had more sense before.